Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt

```diff
@@ -48,4 +55,71 @@ class SimperiumTagsRepository(
             false
         }
     }
+
+    override suspend fun deleteTag(tag: Tag) {
+        deleteTagFromNotes(tag)
+        tag.delete()
+    }
+
+    private fun deleteTagFromNotes(tag: Tag) {
+        val cursor = tag.findNotes(notesBucket, tag.name)
+
+        while (cursor.moveToNext()) {
+            val note = cursor.getObject()
+            val tags = note.tags
+            tags.remove(tag.name)
```

Comment:
we need to keep in mind the distinction here between a tag's name and its `tagHash`

tags are stored in a note's model with the name, of which there might be multiple "lexical variations" that collapse to the same reference, or hash. for example, tags with different casings are identical as are tags whose only difference is their Unicode normal forms.

it would be nice, especially since we're jumping in between notes and tags here, coupling the two, to add a new method to `Note`, which would remove a tag and maintain all the appropriate logic.

```kt
while (cursor.moveToNext()) {
    var note = cursor.getObject()
    note.removeTag(tag.name)
    note.save()
}
```

```kt
// Note
fun removeTag(tag: Tag) {
    tags = tags.filter { name -> ! tag.sameAs(name) }
}
```

Reply:
Fixed in 0429e79402759be2ba4fd93d060e03173f8126bf. I noticed that when the tag is added, it adds the canonical form. For example, if `tag3` exists and I type `TAG3`, the app changes to `tag3`.

Reply:
> it adds the canonical form. For example, if tag3 exists and I type TAG3, the app changes to tag3.

we've still been settling these decisions. the goal is to be able to prompt someone that what they are adding as a new tag already exists, and then to give them a choice to rename the existing tag or adopt the existing name.

we did decide that in the individual note view, when listing tags, that those should eventually all shows the canonical name as provided in the `Tag` entity itself, regardless of how it's written in the `Note`'s `tags` property.

Reply:
There is also the case of two different users adding lexical variations of the same tag before synchronizing. For example, user 1 adds `TAG3` and user 2 adds `tag3`. The note will keep the two tags I think.